api/v1alpha1/gitrepository_types.go

```diff
@@ -0,0 +1,75 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type GitRepositorySpec struct {
+	Source GitRepositorySource `json:"source,omitempty"`
+	// GitURL is the base URL of Git server used for API calls.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=`^https?:\/\/.+$`
+	GitURL string `json:"gitURL"`
+	// SecretRef is the reference to secret that contain Git server credentials
+	// +kubebuilder:validation:Optional
+	SecretRef SecretReference `json:"secretRef"`
+}
+
+type GitRepositorySource struct {
+	// +kubebuilder:validation:Enum:=argocd;backstage;crossplane;gitea;nginx
+	// +kubebuilder:validation:Optional
+	EmbeddedAppName string `json:"embeddedAppName"`
+	// Path is the absolute path to directory that contains Kustomize structure or raw manifests.
+	// This is required when Type is set to local.
+	// +kubebuilder:validation:Optional
+	Path string `json:"path"`
+	// Type is the source type.
+	// +kubebuilder:validation:Enum:=local;embedded
+	// +kubebuilder:default:=embedded
+	Type string `json:"type"`
+}
+
+type SecretReference struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+}
+
+type Commit struct {
+	// Hash is the digest of the most recent commit
+	// +kubebuilder:validation:Optional
+	Hash string `json:"hash"`
+}
+
+type GitRepositoryStatus struct {
+	// LatestCommit is the most recent commit known to the controller
+	// +kubebuilder:validation:Optional
+	LatestCommit Commit `json:"commit"`
+	// ExternalGitRepositoryUrl is the url for the in-cluster repository accessible from local machine.
+	// +kubebuilder:validation:Optional
+	ExternalGitRepositoryUrl string `json:"externalGitRepositoryUrl"`
+	// GitRepositoryUrl is the url for the in-cluster repository accessible within the cluster.
+	// +kubebuilder:validation:Optional
+	GitRepositoryUrl string `json:"gitRepositoryUrl"`
```

Comment:
Why don't you name it InternalGitRepositoryUrl as it is used for internal purposes ? If we use gitea which is accessible by the users, does that makes sense that we keep ext vs internal repos ?

Reply:
Good idea I will rename the field.

We need to distinction because the external and internal git repo url (`gitea.cnoe.localtest.me`) because the domain name will always resolve to 127.0.0.1. Therefore this URL is not usable by ArgoCD to fetch manifests from the gitea server.